Refactor ProbeResultNormalizer

Maxr1998 opened the pull request:
**Changes**
Improves code structure and readability.
I've had this lying around for a while, and since #5696 is merged, I can open a PR now.

cvium commented on MediaBrowser.MediaEncoding/Probing/FFProbeHelpers.cs:
```diff
@@ -57,6 +57,31 @@ public static string GetDictionaryValue(IReadOnlyDictionary<string, string> tags
+        /// <param name="tags">The tags.</param>
+        /// <param name="keys">The checked keys.</param>
+        /// <returns>System.String.</returns>
+        public static string GetFirstNonNullNorBlankDictionaryValue(IReadOnlyDictionary<string, string> tags, params string[] keys)
```
The problem with this is that it allocates an array every time you call it

Maxr1998 replied:
Ah, so it's similar to Java where vararg creates an implicit copy, on top of the array declared at the callsite?

cvium replied:
Well, you haven't declared an array per se. You're just letting it convert the arguments to an array at runtime.

Maxr1998 replied:
Oh, then I meant something different.
Regardless, having a huge block of if checks is quite verbose, that's why I introduced that helper function. I could of course switch the `params` to just using an array, and declare the array with the dictionary keys statically in the file and not in the function.
That way, it creates less allocations when called multiple times. Would that work better?

cvium replied:
The options I see are

1. Create multiple methods with varying number of arguments
2. Create specialized methods for every parameter `GetPremiereDate(tags)`

Maxr1998 replied:
I created a [small benchmark](https://gist.github.com/Maxr1998/d5ea87ee42d8076311f835865f6da1c5) to test various alternatives. In all cases, the worst case of no results in the dictionary was used to test.

As expected, the array allocation create a huge performance hit, so that option should definitely be avoided. Declaring the array as class attribute, thus shared across allocations, performs quite well, but the overall memory usage of the program would increase, so maybe not the best option.
Using a hardcoded argument count performs best, so that's probably the route to take. I propose supporting 1, 2 and 3 parameters for the function. Specialized methods for each of the attributes (name, album artist, etc.) is imo a bit too verbose/specific.

Maxr1998 replied:
I implemented the suggested changes in 3e9570a6f55fff97724e5527a8fb39703b09d542. Should be good now.

cvium commented on MediaBrowser.MediaEncoding/Probing/FFProbeHelpers.cs:
```diff
@@ -57,6 +57,90 @@ public static string GetDictionaryValue(IReadOnlyDictionary<string, string> tags
+                return val;
+            }
+
+            if (tags.TryGetValue(key3, out val) && !string.IsNullOrWhiteSpace(val))
```
"a lot" of code duplication here. I think it would be better to create one method that handles 3 keys and simply sending string.Empty from the other two methods.

Eg. `if (!string.IsNullOrEmpty(key3) && tags.TryGetValue(key3, out val) && !string.IsNullOrWhiteSpace(val))`

Maxr1998 replied:
Right, copy pasting those wasn't the greatest decision. And I think so, yes, but the other method should be moved as well then.

cvium replied:
> I think so, yes, but the other method should be moved as well then.

Do you mean GetDictionaryValue? There is already an extension method for this:
[GetValueOrDefault](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.collectionextensions.getvalueordefault?view=net-5.0#System_Collections_Generic_CollectionExtensions_GetValueOrDefault__2_System_Collections_Generic_IReadOnlyDictionary___0___1____0_)

Maxr1998 replied:
> Do you mean GetDictionaryValue? There is already an extension method for this:
> [GetValueOrDefault](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.collectionextensions.getvalueordefault?view=net-5.0#System_Collections_Generic_CollectionExtensions_GetValueOrDefault__2_System_Collections_Generic_IReadOnlyDictionary___0___1____0_). The other methods could be moved though

Addressed this now.

cvium commented on MediaBrowser.Common/Extensions/DictionaryExtensions.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <param name="key2">The second checked key.</param>
+        /// <param name="key3">The third checked key.</param>
+        /// <returns>System.String.</returns>
+        public static string? GetFirstNonNullNorBlankDictionaryValue(this IReadOnlyDictionary<string, string>? dictionary, string key1, string key2, string key3)
```
Been thinking about the name. Maybe `GetFirstNotNullNorEmptyValue`?

Maxr1998 replied:
Right, the `Dictionary` can be removed. I'd keep the `Blank` though or, better, change it to `WhiteSpace` to make it obvious that it doesn't return whitespace.

cvium commented on MediaBrowser.MediaEncoding/Probing/FFProbeHelpers.cs:
```diff
@@ -46,7 +44,7 @@ public static void NormalizeFFProbeResult(InternalMediaInfoResult result)
         /// <param name="key">The key.</param>
         /// <returns>System.String.</returns>
-        public static string GetDictionaryValue(IReadOnlyDictionary<string, string> tags, string key)
+        public static string? GetDictionaryValue(IReadOnlyDictionary<string, string>? tags, string key)
```
Is `tags` ever null? And shouldn't this method be replaced with `GetValueOrDefault`?

Maxr1998 replied:
It shouldn't, but Rider warned that it might be called with `null` values, and should thus be checked. They suggested to use a contract or throwing an exception, I just kept the current behavior of handling that value.

Maxr1998 replied:
Re edit: so just remove the check and nullability and trust that it won't break?

Maxr1998 commented on MediaBrowser.Common/Extensions/DictionaryExtensions.cs:
```diff
@@ -0,0 +1,69 @@
+                return val;
+            }
+
+            if (dictionary.TryGetValue(key2, out val) && !string.IsNullOrWhiteSpace(val))
```
I somehow missed this. Thanks, will fix.

Maxr1998 commented:
I rebased/squashed the commits onto the latest master and removed the DictionaryExtensions to Jellyfin.Extensions. Also addressed the review feedback (added the parameter empty/null checks, renamed the helper functions).

cvium commented on MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:
```diff
@@ -923,17 +873,13 @@ private string GetDictionaryValue(IReadOnlyDictionary<string, string> tags, stri
-            }
-
-            return input.Split('(').FirstOrDefault();
+            return !string.IsNullOrEmpty(input) ? input.Split('(').FirstOrDefault() : input;
```
I don't think this is an improvement

cvium commented on MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:
```diff
@@ -1018,117 +964,119 @@ private bool IsClose(double d1, double d2, double variance = .005)
+                return;
             }
+
+            info.Size = !string.IsNullOrEmpty(data.Format.Size)
```
I generally prefer not having negations. It flows better in my mind if I don't have to negate an expression.

Maxr1998 replied:
Agreed, unfortunately there's no inverse method for this I fear.

cvium replied:
You could just not negate it :P

```C#
             info.Size = string.IsNullOrEmpty(data.Format.Size)
                ? null
                : long.Parse(data.Format.Size, _usCulture);
```

Maxr1998 replied:
Right, but I personally think the common/normal case should be handled first, will fallback/error cases second.

cvium commented on MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:
```diff
@@ -1167,69 +1115,49 @@ private void SetAudioInfoFromTags(MediaInfo audio, Dictionary<string, string> ta
-                    .ToArray();
-            }
+            // Set album artist
+            var albumArtist = tags.GetFirstNotNullNorWhiteSpaceValue("album_artist", "albumartist", "album artist");
```
Is the change in ordering intentional? To match the previous behaviour it should be this
```suggestion
            var albumArtist = tags.GetFirstNotNullNorWhiteSpaceValue("albumartist", "album artist", "album_artist");
```

Maxr1998 replied:
It's not, I'll revert it to the previous behavior.

cvium commented on MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:
```diff
@@ -1531,24 +1449,27 @@ private void FetchWtvInfo(MediaInfo video, InternalMediaInfoResult data)
-                        video.Timestamp = null;
-                    }
-                }
+            if (!string.Equals(video.Container, "mpeg2ts", StringComparison.OrdinalIgnoreCase) &&
+                !string.Equals(video.Container, "m2ts", StringComparison.OrdinalIgnoreCase) &&
+                !string.Equals(video.Container, "ts", StringComparison.OrdinalIgnoreCase))
```
```suggestion
            if (!string.Equals(video.Container, "mpeg2ts", StringComparison.OrdinalIgnoreCase)
                && !string.Equals(video.Container, "m2ts", StringComparison.OrdinalIgnoreCase)
                && !string.Equals(video.Container, "ts", StringComparison.OrdinalIgnoreCase))
```

Maxr1998 replied:
Just for the future, is there a code style linter for stuff like this? E.g. in Kotlin, `detekt` wants you to put logical operators on the previous line. Here it seems to be the other way around.

cvium replied:
Not that I'm aware of

Bond-009 commented on MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:
```diff
@@ -1299,95 +1227,90 @@ private IEnumerable<string> SplitArtists(string val, char[] delimiters, bool spl
             }
+
+            info.Studios = studioList
+                .Where(i => !string.IsNullOrWhiteSpace(i))
```
Could do this at the beginning of the foreach loop, no?

Maxr1998 replied:
Yep. Changed it. Instead of the `Distinct`, we could event change it to use a Set, right?

Bond-009 commented on MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:
```diff
@@ -1299,95 +1227,90 @@ private IEnumerable<string> SplitArtists(string val, char[] delimiters, bool spl
             }
+
+            info.Genres = genres
+                .Where(i => !string.IsNullOrWhiteSpace(i))
```
Same

Bond-009 commented on MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:
```diff
@@ -1442,29 +1367,23 @@ private void FetchWtvInfo(MediaInfo video, InternalMediaInfoResult data)
 
-            var year = FFProbeHelpers.GetDictionaryValue(data.Format.Tags, "WM/OriginalReleaseTime");
-            if (!string.IsNullOrWhiteSpace(year))
+            var year = tags.GetValueOrDefault("WM/OriginalReleaseTime");
```
TryGetValue?

Maxr1998 commented on MediaBrowser.MediaEncoding/Probing/FFProbeHelpers.cs:
```diff
@@ -41,38 +39,23 @@ public static void NormalizeFFProbeResult(InternalMediaInfoResult result)
+        public static int? GetDictionaryNumericValue(IReadOnlyDictionary<string, string> tags, string key)
         {
-            if (tags == null)
+            var val = tags.GetValueOrDefault(key);
+
+            if (string.IsNullOrEmpty(val))
```
@cvium suggested `GetValueOrDefault` :stuck_out_tongue:

Would be this then, right?
```C#
if (!tags.TryGetValue(key, out var val) || string.IsNullOrEmpty(val))
{
    return null;
}
```

What's the problem with `GetValueOrDefault`?

Maxr1998 replied:
Nvm, looked at the source, it simply calls `TryGetValue` anyway, just with more overhead. Gonna change it where applicable.
I'd argue that in the ProbeResultNormalizer it'd be preferable to keep the code concise.

Maxr1998 replied:
Gave it another look and changed it, should be more readable now. Also removed the `isNullOrEmpty` check because that fails the conversion anyway.

Maxr1998 commented on MediaBrowser.MediaEncoding/Probing/FFProbeHelpers.cs:
```diff
@@ -84,9 +67,9 @@ public static string GetDictionaryValue(IReadOnlyDictionary<string, string> tags
+        public static DateTime? GetDictionaryDateTime(IReadOnlyDictionary<string, string> tags, string key)
         {
-            var val = GetDictionaryValue(tags, key);
+            var val = tags.GetValueOrDefault(key);
```
Fixed, see above.

Maxr1998 commented on MediaBrowser.MediaEncoding/Probing/FFProbeHelpers.cs:
```diff
@@ -41,38 +39,21 @@ public static void NormalizeFFProbeResult(InternalMediaInfoResult result)
             if (tags == null)
             {
-                return null;
+                throw new ArgumentNullException(nameof(tags));
```
It should never be null. We could simply leave out the check, but then the linter complains. An explicit throw is best practice imo.

Maxr1998 commented on MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:
```diff
@@ -1442,29 +1374,23 @@ private void FetchWtvInfo(MediaInfo video, InternalMediaInfoResult data)
 
-            var premiereDateString = FFProbeHelpers.GetDictionaryValue(data.Format.Tags, "WM/MediaOriginalBroadcastDateTime");
-            if (!string.IsNullOrWhiteSpace(premiereDateString))
+            var premiereDateString = tags.GetValueOrDefault("WM/MediaOriginalBroadcastDateTime");
```
Done, another thing even - when parsing, it uses the year, not the premiereDateString - is that a bug?
